### ISSUE TITLE
NBA LEAGUE Streams not working 

### ISSUE THREAD

**Issue by crak3rjack** [d21spike/plugin.video.sling#40]
@d21spike Regular channels are working fine, but all NBA LEAGUE PASS streams are not working... here's the log and thanks
https://pastebin.com/puUCB4Up

**Comment by crak3rjack** [comment]
@d21spike @eracknaphobia Is it possible to get a fix👆

**Comment by d21spike** [comment]
Not ignoring you. I glanced at your log, mechanically the addon is working like it should but Sling is responding with 404 - Not Found when the stream is actually played. Being that these channels aren't always live, it becomes one of timing in order to observe more deeply.

**Comment by crak3rjack** [comment]
@d21spike So you need more time to look at this error from what i can understand.

**Comment by d21spike** [comment]
Ok so I did some more looking, it looks to be that only the channels with a 4 hour time slot but no guide data are the ones failing playback with an error similar to this:

``` python
ERROR <general>: AddOnLog: inputstream.adaptive: Download failed with error 404: http://p-cdn3-805-cg14-linear-cbd46b77.movetv.com/15805/live/NBADET/32276cd63a3111ecbf660025b5471213/video/vid03/00006767.m4s
```

That said this is in a portion that is handled by IA and not this add-on. This is not to blame IA but basically currently there isn't anything to playback so it fails with a 404 Not Found, which is technically correct since like I said there is no guide data. From what I can tell right now the channels are: ```Brooklyn Nets, Charlotte Hornets, Detroit Pistons, and Portland Trail Blazers```.
![image](https://user-images.githubusercontent.com/6399602/139608236-d8427378-b826-4289-9991-dd3798dd0649.png)

**I would like to know what the behavior is when there is actually a game on. Can you tell me if it plays on Sling via the browser? Can you take a picture of what the guide looks like when there is a game on?**

The other NBA channels have Off-Air for the playback as well as guide. This basically means trying to watch the channel will result in a screen similar to this:
![image](https://user-images.githubusercontent.com/6399602/139608395-53dbc49c-7fb7-4bc9-b324-b7ccedd64f26.png)

Today I pushed an update. This adds a setting to the add-on, disabled by default. Enabling it will display Off-Air channels in the ```Channels Menu```.
![image](https://user-images.githubusercontent.com/6399602/139608443-84d6264b-d19c-44af-8ec3-922873bc3601.png)

![image](https://user-images.githubusercontent.com/6399602/139608456-207e3d3f-1c41-4fcf-a059-ed03e6928f55.png)

As you can see, majority of these channels have a 1-hour time slot and play no problem. Either way try the setting and see if you can get the channels to play. Also see if you can answer the couple questions from above.

**Comment by crak3rjack** [comment]
@d21spike Im using the latest update you just pushed and still not working also enabled the off-air channel setting and nothing works in the addon... but it does work on the website.
![1](https://user-images.githubusercontent.com/59179646/139610481-7c77b90b-c7ea-4e41-a0c1-83f92978835a.jpg)
![2](https://user-images.githubusercontent.com/59179646/139610489-3fd35280-61b4-4ce3-b4fb-c8a9c73654df.jpg)
![3](https://user-images.githubusercontent.com/59179646/139610495-07e7ef9c-e656-4774-bb90-b2030df9011c.jpg)

g

**Comment by d21spike** [comment]
Ok now we're getting somewhere. Try playing the ```Brooklyn Nets``` channel via Sling Addon (not Sling Guide my recent update doesn't affect this). This is what I'm seeing,...
![image](https://user-images.githubusercontent.com/6399602/139610898-4ed966ae-b2bc-42c2-b36e-b21cd6d0cd6f.png)

**Comment by crak3rjack** [comment]
@d21spike just tried in the addon On Now > Sports > Nets... and those work. Takes a while to load.

**Comment by d21spike** [comment]
This is what that Setting currently affects

You would toggle this to ```On```
![image](https://user-images.githubusercontent.com/6399602/139611213-402d00d3-ab09-4701-9fdb-b7b0c722d4b0.png)

Then ```Sling```
![image](https://user-images.githubusercontent.com/6399602/139611261-7a44b2c2-8755-4035-869b-1c0147486898.png)

Then ```Channels```
![image](https://user-images.githubusercontent.com/6399602/139611300-f5a0bc64-c659-4239-8816-7e3025eff6cc.png)

Then ```Brooklyn Nets```
![image](https://user-images.githubusercontent.com/6399602/139611420-8d11aaa7-7a3a-4229-aef9-75bc4177cc93.png)

**Comment by d21spike** [comment]
Or was it that you were just having trouble playing via the ```Sling Guide```?

**Comment by crak3rjack** [comment]
@d21spike  It's working in the addon, thats good enough for me im super happy... hopefully you can get it to work in the guide "not working" next and appreciate the quick response and help bud!☺👍

**Comment by d21spike** [comment]
@crak3rjack Glad to hear it's working. Just to address the ```Guide``` ... technically it **does** work. See mine below for example, it does play the Nets since there is something on the "Programme". When the schedule says ```Off-Air``` or ```Nothing Scheduled``` I have it coded not to even try play since technically there isn't anything on anyways.
![image](https://user-images.githubusercontent.com/6399602/139611935-d8612100-4bd5-4b38-a756-d5f55a95d430.png)

Looking at your screenshot above
![image](https://user-images.githubusercontent.com/6399602/139611966-271f01ec-73d4-4cd0-900a-1bf036b0d495.png)

You're working with an incomplete guide. This basically tells me two things. 

1. You don't use the Slinger service
2. You haven't visited the ```Channels``` menu in around 24 hours

Basically the guide relies on the database that the Sling addon maintains. The channels specifically and their guide data are updated either via the Slinger service automatically (IF ENABLED and you can choose how many days you'd like) or every time you visit the channels menu, 24 hours of Guide data is downloaded. Usually 12 am to 12 am.

The basic gist is, the Guide is quick since it doesn't have to retrieve any data it's all stored on your device. If I were to add the update to the Guide it would create a substantial slow down. So my suggestion would be either to visit the ```Channels``` menu daily or enable the ```Slinger Service``` and have it just run for ```Channels``` and ```Guide```. This will ensure you have data when needed.

**Comment by crak3rjack** [comment]
@d21spike For reason tonight im getting this error on all the games that are on.
https://pastebin.com/K55meQy7

**Comment by d21spike** [comment]
Looks like IA is having trouble decrypting. Gonna have to take a deeper look into it.
```python
2021-11-01 20:34:10.130 T:3212    DEBUG <general>: AddOnLog: inputstream.adaptive: CDMMessage: 2 arrived!
2021-11-01 20:34:10.138 T:3212    DEBUG <general>: AddOnLog: inputstream.adaptive: Sessionkey:780182A9AA5F4EF2863B685768A78417 status: 0 syscode: 0
2021-11-01 20:34:10.138 T:3212    DEBUG <general>: AddOnLog: inputstream.adaptive: CDMMessage: 3 arrived!
2021-11-01 20:34:10.142 T:3212    DEBUG <general>: AddOnLog: inputstream.adaptive: License update successful
2021-11-01 20:34:10.158 T:3212    DEBUG <general>: AddOnLog: inputstream.adaptive: DecryptSampleData: Decrypt failed with error: 2 and key: 780182A9AA5F4EF2863B685768A78417
2021-11-01 20:34:10.159 T:3212    DEBUG <general>: AddOnLog: inputstream.adaptive: Initializing stream with KID: 780182A9AA5F4EF2863B685768A78417
2021-11-01 20:34:10.162 T:3212    DEBUG <general>: AddOnLog: inputstream.adaptive: DecryptSampleData: Decrypt failed with error: 2 and key: 780182A9AA5F4EF2863B685768A78417
2021-11-01 20:34:10.162 T:3212    DEBUG <general>: AddOnLog: inputstream.adaptive: GetCapabilities()
```

At @matthuisman any known issues with IA or you thinking this may be a licensing issue? More complete log below
[PASTEBIN](https://pastebin.com/0DdSMtTB)

**Comment by matthuisman** [comment]
that Decrypt failed error is normal.
your actual error is 
```
2021-11-01 20:34:10.288 T:3212    DEBUG <general>: AddOnLog: inputstream.adaptive: Build segments from SIDX atom...
2021-11-01 20:34:10.288 T:3212    ERROR <general>: AddOnLog: inputstream.adaptive: Unable to create SIDX from IndexRange bytes
```
you'll need to open an issue on inputstream adaptive.
Youll need to provide the manifest and full debug log etc.
There is an issue template there to follow

I suspect it may be the same issue as mine here:
https://github.com/xbmc/inputstream.adaptive/issues/731
Without a manifest, it's hard to tell

i workaround the above issue inside my proxy thats uses for all my requests in and out of my add-on with the below code
https://github.com/matthuisman/slyguy.addons/blob/56c417ec193d60028f0f23d3ede055727626e1b7/script.module.slyguy/resources/lib/proxy.py#L369

basically if timeShiftBufferDepth is not in the mpd attributes, I create and add it

**Comment by crak3rjack** [comment]
@d21spike @matthuisman Is there any work around for the issue... inputstream takes a lot time for fixes & thanks.

**Comment by matthuisman** [comment]
@crak3rjack 
already answered above your comment...
without a manifest - dont know if its same issue.
If it is same issue - then i've worked around it using a proxy.

**Comment by crak3rjack** [comment]
@matthuisman Noob here can you explain me on how to add the proxy into Sling to give it a try if it's not to tech and thanks.

**Comment by matthuisman** [comment]
no sorry, that is not easy at all to do.
what you can do to help is download the manifest (url) that doesnt play and upload it here

**Comment by crak3rjack** [comment]
@matthuisman Sorry only know how get a log file don't even know how a manifest :(

**Comment by matthuisman** [comment]
in the log, you will see the ulr of the file its trying to play
eg. http://p-cdn1-802-cg14-linear-cbd46b77.movetv.com/clipslist/1005/20211101T233000Z/20211102T033000Z.mpd
copy this url into browser and it will download that file.
Then upload it here

**Comment by crak3rjack** [comment]
@matthuisman there's only this for video & audio
<general>: AddOnLog: inputstream.adaptive: Download failed with error 404: http://p-cdn3-802-cg14-linear-cbd46b77.movetv.com/15802/live/NBACLE/aa69bafa4d8211ecbc620025b5472112/video/vid03/ffffe88d.m4s

**Comment by matthuisman** [comment]
either before or after that in the log, there should be an entry telling you the manifest url.
It will end in ".mpd"
find that url, and then download it in browser straight away before it expires.
Then upload it

**Comment by crak3rjack** [comment]
@d21spike Now that got login im getting this error in all regular streams
2021-12-04 17:36:17.914 T:12976    INFO <general>: VideoPlayer::OpenFile: plugin://plugin.video.sling/?url=http%3A%2F%2Fcbd46b77.cdn.cms.movetv.com%2Fcms%2Fapi%2Fchannels%2Fbe5a3bce259a4c13947f4b4fbae5e005%2Fschedule%2Fnow%2Fplayback_info.qvt&mode=play&name=A%26E
2021-12-04 17:36:17.915 T:17496    INFO <general>: Creating InputStream
2021-12-04 17:36:17.917 T:17496    INFO <general>: AddOnLog: inputstream.adaptive: SetVideoResolution (1920 x 1080)
2021-12-04 17:36:18.295 T:17496    INFO <general>: AddOnLog: inputstream.adaptive: Successfully parsed manifest file. #Periods: 1, #Streams in first period: 2, Type: live, Download speed: 817668.6014 Bytes/s
2021-12-04 17:36:18.991 T:17496    INFO <general>: Creating Demuxer
2021-12-04 17:36:18.992 T:17496    INFO <general>: Opening stream: 1001 source: 256
2021-12-04 17:36:19.244 T:17496    INFO <general>: Creating video codec with codec id: 27
2021-12-04 17:36:19.244 T:17496    INFO <general>: CDVDVideoCodecFFmpeg::Open() Using codec: H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
2021-12-04 17:36:19.245 T:17496    INFO <general>: Creating video thread
2021-12-04 17:36:19.245 T:13068    INFO <general>: running thread: video_thread
2021-12-04 17:36:19.246 T:17496    INFO <general>: Opening stream: 1002 source: 256
2021-12-04 17:36:19.327 T:17496    INFO <general>: Finding audio codec for: 86018
2021-12-04 17:36:19.328 T:17496    INFO <general>: CDVDAudioCodecFFmpeg::Open() Successful opened audio decoder aac
2021-12-04 17:36:19.328 T:17496    INFO <general>: Creating audio thread
2021-12-04 17:36:19.328 T:19740    INFO <general>: running thread: CVideoPlayerAudio::Process()
2021-12-04 17:36:19.546 T:19744   ERROR <general>: AddOnLog: inputstream.adaptive: Download failed with error 404: http://p-cdn3-805-cg14-linear-cbd46b77.movetv.com/15805/live/AESTR/de7441ba54e611ecaa790025b5471114/video/vid06/00005839.m4s
2021-12-04 17:36:30.123 T:19744   ERROR <general>: Skipped 8 duplicate messages..
2021-12-04 17:36:30.123 T:19744   ERROR <general>: AddOnLog: inputstream.adaptive: Download failed with error 404: http://p-cdn3-805-cg14-linear-cbd46b77.movetv.com/15805/live/AESTR/de7441ba54e611ecaa790025b5471114/video/vid06/00005839.m4s
2021-12-04 17:36:31.571 T:16748   ERROR <general>: Skipped 1 duplicate messages..
2021-12-04 17:36:31.571 T:16748   ERROR <general>: AddOnLog: inputstream.adaptive: Download failed with error 404: http://p-cdn3-805-cg14-linear-cbd46b77.movetv.com/15805/live/AESTR/de7441ba54e611ecaa790025b5471114/audio/stereo/192/00005839.m4s
2021-12-04 17:36:42.119 T:16748   ERROR <general>: Skipped 7 duplicate messages..
2021-12-04 17:36:42.119 T:16748   ERROR <general>: AddOnLog: inputstream.adaptive: Download failed with error 404: http://p-cdn3-805-cg14-linear-cbd46b77.movetv.com/15805/live/AESTR/de7441ba54e611ecaa790025b5471114/audio/stereo/192/00005839.m4s
2021-12-04 17:36:44.549 T:17496    INFO <general>: Skipped 2 duplicate messages..
2021-12-04 17:36:44.549 T:17496    INFO <general>: CVideoPlayer::Process - eof reading from demuxer
2021-12-04 17:36:44.549 T:17496    INFO <general>: CVideoPlayer::OnExit()
2021-12-04 17:36:44.549 T:17496    INFO <general>: VideoPlayer: eof, waiting for queues to empty
2021-12-04 17:36:44.549 T:17496    INFO <general>: Closing stream player 1
2021-12-04 17:36:44.550 T:17496    INFO <general>: Waiting for audio thread to exit
2021-12-04 17:36:44.554 T:19740    INFO <general>: thread end: CVideoPlayerAudio::OnExit()
2021-12-04 17:36:44.555 T:17496    INFO <general>: Closing audio device
2021-12-04 17:36:44.555 T:17496    INFO <general>: Deleting audio codec
2021-12-04 17:36:44.555 T:17496    INFO <general>: Closing stream player 2
2021-12-04 17:36:44.555 T:17496    INFO <general>: waiting for video thread to exit
2021-12-04 17:36:44.555 T:13068   ERROR <general>: Got MSGQ_ABORT or MSGO_IS_ERROR return true
2021-12-04 17:36:44.555 T:13068    INFO <general>: thread end: video_thread
2021-12-04 17:36:44.555 T:17496    INFO <general>: deleting video codec
2021-12-04 17:36:44.559 T:14512    INFO <general>: Deleting settings information for files plugin://plugin.video.sling/?url=http%3A%2F%2Fcbd46b77.cdn.cms.movetv.com%2Fcms%2Fapi%2Fchannels%2Fbe5a3bce259a4c13947f4b4fbae5e005%2Fschedule%2Fnow%2Fplayback_info.qvt&mode=play&name=A%26E
2021-12-04 17:36:44.565 T:17496    INFO <general>: ADDON: Dll Destroyed - InputStream Adaptive
2021-12-04 17:36:44.575 T:12976    INFO <general>: CVideoPlayer::CloseFile()
2021-12-04 17:36:44.576 T:12976    INFO <general>: VideoPlayer: waiting for threads to exit
2021-12-04 17:36:44.576 T:12976    INFO <general>: VideoPlayer: finished waiting
2021-12-04 17:36:44.576 T:12976    INFO <general>: CVideoPlayer::CloseFile()
2021-12-04 17:36:44.576 T:12976    INFO <general>: VideoPlayer: waiting for threads to exit
2021-12-04 17:36:44.576 T:12976    INFO <general>: VideoPlayer: finished waiting
2021-12-04 17:36:46.398 T:13760 WARNING <general>: Attempt to use invalid handle 9
2021-12-04 17:36:46.398 T:13760    INFO <general>: Skipped 5 duplicate messages..
2021-12-04 17:36:46.398 T:13760    INFO <general>: CPythonInvoker(32, C:\Users\jrh\AppData\Roaming\Kodi\addons\plugin.video.sling\default.py): script successfully run
2021-12-04 17:36:46.596 T:13760    INFO <general>: Python interpreter stopped

**Comment by matthuisman** [comment]
a user has provided an example mpd dash
[20211210T010000Z.mpd.txt](https://github.com/d21spike/plugin.video.sling/files/7690789/20211210T010000Z.mpd.txt)

as you will see, there is no timeShiftBufferDepth but there is mediaPresentationDuration 
So, i dont think its the same issue as https://github.com/xbmc/inputstream.adaptive/issues/731

Looking more at it, Inputstream will be treating it as VOD (not live) due to type="static"
However, it also has profiles="urn:mpeg:dash:profile:isoff-live:2011"

But lots of VOD stuff uses profiles="urn:mpeg:dash:profile:isoff-live:2011"
Basically it just means it was recorded off a live feed.

Maybe the MPD sent to me was actually VOD and not LIVE?

**Comment by crak3rjack** [comment]
@d21spike Hi bud is there gonna be a fix for the NBA LEAGUE PASS and thanks!

**Comment by eracknaphobia** [comment]
@matthuisman Here is a live one. Looks like you are right assuming it's trying to play a segment that is out of range.
[spanning_ads.mpd.txt](https://github.com/d21spike/plugin.video.sling/files/7901326/spanning_ads.mpd.txt)

The .m4s urls it's trying to call return json
http://p-cdn3-802-cg14-linear-cbd46b77.movetv.com/15802/live/NBABRK/f1d492c0798311ec9d770025b5472213/video/vid04/ffffe556.m4s
```
{
	"status": 404,
	"status_string": "Not Found",
	"message": "Request for stream f1d492c0798311ec9d770025b5472213 segment 0xffffe556 is beyond the stream end (0x2c0; limit 0x2c1)"
}

```

**Comment by matthuisman** [comment]
@eracknaphobia 
could you try a build from here:
https://jenkins.kodi.tv/blue/organizations/jenkins/xbmc%2Finputstream.adaptive/detail/PR-879/1/artifacts
(for Matrix)

**Comment by eracknaphobia** [comment]
That did allow the stream to play. However, it started the stream from the beginning instead of "live".

**Comment by matthuisman** [comment]
if the stream is live and multiple-periods - then you need to fast fwd to live.
Or, in the add-on use a large resume_from to force live.
see here: https://github.com/xbmc/inputstream.adaptive/issues/574

**Comment by eracknaphobia** [comment]
@matthuisman With the latest build I'm noticing the live channels will play from beginning when the ResumeTime and TotalTime are set to 1, but when I try `li.setProperty('ResumeTime', '43200')` to play "live" nothing plays and I see the 404's for the m4s files.
[spanning_ads.mpd.txt](https://github.com/d21spike/plugin.video.sling/files/7906842/spanning_ads.mpd.txt)

**Comment by matthuisman** [comment]
My build was just a hack to try something. it would require more investigation which I don't have time for currently sorry.

It seems IA is calculating the wrong length of the stream. 

What would help is I'd someone opens a new issue in IA and provide

1) play the channel in browser sling website and find the segment number it is using when playing at the live head
2) download the mpd manifest as quickly as possible
3) record the current epoch time

It won't be exact but would give a pretty good target of what IA should be calculating

All that is possible using chrome webtools. That records the time of requests and will have the manifest data too

The calculation I think should be something like

1) find how many seconds since the start of the stream (time - availability start)
Let's say 1000 seconds
(This should be the total length that can be navigated in Kodi if no buffer length advised)
2) minus the start time of the current period 
Eg. Current period started at 100s. That gives us 900s length of current period.
3) duration of segment / timescale 2048/1000 = 2.48second per segment
4) 900s / 2.48 = 362
5) live head segment = #362 + start number
https://docs.aws.amazon.com/mediapackage/latest/ug/segtemp-format-duration.html

Looking at the IA code, looks like we doing timescale/duration which is wrong. 

Also adding 1 to account for rounding should probably actually minus 1. Better to be grabbing a segment 1 behind head than 1 ahead (404)

And I think the oversalsecs is used incorrectly / shouldn't be used in this calculation

**Comment by crak3rjack** [comment]
@eracknaphobia @matthuisman Appreciate it guys for taking of your time to look into the issue thanks!🙂💪

**Comment by crak3rjack** [comment]
@eracknaphobia @matthuisman I've seen you guys testing the issue on Windows PC, I tried the inputstream from @matthuisman in Android last night and didn't work can you guys take a look and confirm it works also in Android and thanks.

**Comment by eracknaphobia** [comment]
I just installed the android version. I will test some of the games tonight.

**Comment by crak3rjack** [comment]
@eracknaphobia I tried the arm7 version!🤞👍

**Comment by eracknaphobia** [comment]
@crak3rjack Did the nba streams play from beginning or not at all? Do channels still work for you?

**Comment by crak3rjack** [comment]
@eracknaphobia All regular channels works fine but the NBA makes likes is going to load but keeps spinning doesn't load at all.

**Comment by crak3rjack** [comment]
@eracknaphobia Forgot to tell you that using the inputstream from the oficial Kodi 19.0.2 regular channels don't play in Android.

**Comment by eracknaphobia** [comment]
@crak3rjack oh ok. The version of IA @matthuisman posted [here ](https://github.com/xbmc/inputstream.adaptive/issues/880#issuecomment-1018037719)should play the games from beginning and you can ffwd to live. There will have to be code added to the sling addon to make them play live on start. Still working on that part.

**Comment by matthuisman** [comment]
Test builds here
https://jenkins.kodi.tv/blue/organizations/jenkins/xbmc%2Finputstream.adaptive/detail/PR-881/5/artifacts

**Comment by eracknaphobia** [comment]
@crak3rjack This version of sling should fix the issue. Please test it out and let me know.
[plugin.video.sling-2022.1.24.zip](https://github.com/d21spike/plugin.video.sling/files/7963609/plugin.video.sling-2022.1.24.zip)

**Comment by crak3rjack** [comment]
@eracknaphobia  I'll give it test drive and report back ASPS and hopefully is resolved and thanks!🙏

**Comment by crak3rjack** [comment]
@eracknaphobia @matthuisman @d21spike 
Yes NBA LEAGUE PASS is working "Android" now Celtics vs Pelicans & Pacers vs Mavericks... appreciate all the time and effort. Time to add it to the repo and thanks!😊🙏

**Comment by matthuisman** [comment]
My changes I had done were actually wrong and would break other streams. I'm working on a new PR that should sort it. Will need users to test again when the time comes

**Comment by eracknaphobia** [comment]
@matthuisman I noticed that the ResumeTime hack seemed to work on stock 19.0.2 on macOS if that makes a difference. Thanks for the heads up, happy to test when builds are ready.

**Comment by matthuisman** [comment]
The resume time hack will be needed regardless

**Comment by crak3rjack** [comment]
@eracknaphobia @matthuisman  Hi now im getting this error only on NBA PASS "Android"  here's the log and thanks!
https://pastebin.com/UfHyGRr8
